jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java

```diff
@@ -0,0 +1,231 @@
+import org.eclipse.jetty.webapp.WebAppContext;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class DeploymentTempDirTest
+{
+    private static final Logger LOG = Log.getLogger(DeploymentTempDirTest.class);
+
+    private final WebAppProvider webAppProvider = new WebAppProvider();
+    private final ContextHandlerCollection contexts = new ContextHandlerCollection();
+    private final Path testDir = MavenTestingUtils.getTargetTestingPath(DeploymentTempDirTest.class.getSimpleName());
+    private final Path tmpDir = testDir.resolve("tmpDir");
+    private final Path webapps = testDir.resolve("webapps");
+    private final Server server = new Server();
+
+    @BeforeEach
+    public void setup() throws Exception
+    {
+        ServerConnector connector = new ServerConnector(server);
+        server.addConnector(connector);
+
+        FS.ensureEmpty(testDir);
+        FS.ensureEmpty(tmpDir);
+        FS.ensureEmpty(webapps);
+
+        webAppProvider.setMonitoredDirName(webapps.toString());
+        DeploymentManager deploymentManager = new DeploymentManager();
+        deploymentManager.addAppProvider(webAppProvider);
+        server.addBean(deploymentManager);
+
+        HandlerCollection handlerCollection = new HandlerCollection();
+        handlerCollection.addHandler(contexts);
+        handlerCollection.addHandler(new DefaultHandler());
+        deploymentManager.setContexts(contexts);
+        server.setHandler(handlerCollection);
+    }
+
+    @AfterEach
+    public void stop() throws Exception
+    {
+        server.stop();
+    }
+
+    @Test
+    public void testTmpDirectory() throws Exception
+    {
+        Path warPath = MavenTestingUtils.getTestResourcePath("webapps/foo-webapp-1.war");
+        String deploymentXml = "<Configure class=\"org.eclipse.jetty.webapp.WebAppContext\">\n" +
+            "<Set name=\"war\">" + warPath + "</Set>\n" +
+            "<Set name=\"tempDirectory\">" + tmpDir + "</Set>\n" +
+            "<Set name=\"persistTempDirectory\">false</Set>\n" +
+            "</Configure>";
+
+        server.start();
+        WaitScannerListener listener = new WaitScannerListener();
+        webAppProvider.addScannerListener(listener);
+
+        // Add the webapp xml which will will be detected after scan.
+        listener.reset();
+        createNewFile(webapps, "foo-webapp.xml", deploymentXml);
+        webAppProvider.scan();
+        webAppProvider.scan();
+        listener.future.get(5, TimeUnit.SECONDS);
+        WebAppContext webAppContext = getWebAppContext();
+        assertThat(webAppContext.getTempDirectory(), is(tmpDir.toFile()));
+
+        // Add a known file to the temp directory.
```

Comment:
Put a comment in here that we expect that on a redeploy, as persist == false, we expect the tmp directory to be deleted and re-created, so therefore this marker file should disappear.